### PR TITLE
Fix borrows

### PR DESCRIPTION
**Pull request by burdges**
Incidentally this improve performance slightly because the compiler
avoids some check caused by the extra unsafe code:

Old :
test bench_sha3_256_input_4096_bytes ... bench:      19,252 ns/iter (+/- 2,023) = 212 MB/s
test keccakf_u64                     ... bench:         591 ns/iter (+/- 234) = 42 MB/s

New:
test bench_sha3_256_input_4096_bytes ... bench:      18,195 ns/iter (+/- 851) = 225 MB/s
test keccakf_u64                     ... bench:         574 ns/iter (+/- 19) = 43 MB/s

### CODE REVIEW

**Comment by debris** [comment]
I'm not sure about that... on my machine I see no difference. Have you compared generated assembly?

**Comment by burdges** [comment]
It likely depends upon your rust version, but one should broadly expect unsafe code to grow slower in future as the memory model work proceeds.

File: src/lib.rs [hunk 1]
```diff
@@ -233,6 +219,20 @@ impl Keccak {
         }
     }
 
+    fn a_bytes(&self) -> &[u8] {
```
Comment (rphmeier) [inline src/lib.rs]: I wonder if the compiler could reason better about bounds checks if we returned an `&[u8; PLEN * 8]` here (and likewise in `a_mut_bytes`).

Then the body of each function is just a transmute and we save a pointer's worth of stack space :)

File: src/lib.rs [hunk 2]
```diff
@@ -283,39 +282,38 @@ impl Keccak {
...
-
         let outlen = output.len();
         let rate = self.rate;
 
         // second foldp
         let mut op = 0;
         let mut l = outlen;
         while l >= rate {
-            setout(&a, &mut output[op..], rate);
+            setout(self.a_bytes(), &mut output[op..], rate);
```
Comment (rphmeier) [inline src/lib.rs]: this is actually a pretty important change; I didn't realize it before but `self.a` is aliased here in the prior version

**Comment by burdges** [comment]
I've pushed an update to this branch for the current master and partially addressed #9

**Comment by burdges** [comment]
Oh https://github.com/debris/tiny-keccak/pull/11 is probably better.

**Comment by burdges** [comment]
Now closing since https://github.com/debris/tiny-keccak/pull/11 is merged.

**Pull request closed without merging**